File: algorithms/503. Next Greater Element II/solution.py
class Solution(object):
    def nextGreaterElements(self, nums):
        """
        :type nums: List[int]
        :rtype: List[int]
        """
        s, m = [], [-1] * len(nums)
        for i in range(len(nums)) * 2:
            while s and nums[s[-1]] < nums[i]:
                m[s.pop()] = nums[i]
            s.append(i)
        return m

File: algorithms/503. Next Greater Element II/solution.cc
class Solution {
public:
    vector<int> nextGreaterElements(vector<int>& nums) {
        vector<int> result(nums.size(), -1);
        stack<int> s;
        for ( int i = 0, n = nums.size() ; i < 2 * n; ++i ) {
            while ( s.size() > 0 && nums[s.top()] < nums[i % n] ) {
                result[s.top()] = nums[i % n];
                s.pop();
            }
            s.push(i % n);
        }
        return result;
    }
};
